staging/vhost-device-gpu/src/device.rs

```diff
@@ -710,351 +708,461 @@ impl VhostUserBackend for VhostUserGpuBackend {
+    const CONTROL_QUEUE_DATA_ADDR: GuestAddress = GuestAddress(0x10_000);
+    const CONTROL_QUEUE_SIZE: u16 = 1024;
 
-    fn init() -> (
-        Arc<VhostUserGpuBackend>,
-        GuestMemoryAtomic<GuestMemoryMmap>,
-        VringRwLock,
-    ) {
+    fn init() -> (Arc<VhostUserGpuBackend>, GuestMemoryAtomic<GuestMemoryMmap>) {
         let backend = VhostUserGpuBackend::new(GpuConfig::new(
             SOCKET_PATH.into(),
             GpuMode::ModeVirglRenderer,
         ))
         .unwrap();
         let mem = GuestMemoryAtomic::new(
-            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x1000)]).unwrap(),
+            GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), MEM_SIZE)]).unwrap(),
         );
-        let vring = VringRwLock::new(mem.clone(), 16).unwrap();
-        vring.set_queue_info(0x100, 0x200, 0x300).unwrap();
-        vring.set_queue_ready(true);
 
-        (backend, mem, vring)
+        backend.update_memory(mem.clone()).unwrap();
+        (backend, mem)
     }
 
-    // Prepares a single chain of descriptors
-    fn prepare_descriptors(
-        mut next_addr: u64,
-        mem: &GuestMemoryLoadGuard<GuestMemoryMmap<()>>,
-        buf: &mut Vec<u8>,
-        cmd_type: u32,
-    ) -> Vec<Descriptor> {
-        let mut descriptors = Vec::new();
-        let mut index = 0;
-
-        // Gpu header descriptor
-        let ctrl_hdr = virtio_gpu_ctrl_hdr {
-            type_: cmd_type,
-            ..virtio_gpu_ctrl_hdr::default()
-        };
+    /// Arguments to create a descriptor chain for testing
+    struct TestingDescChainArgs<'a> {
+        readable_desc_bufs: &'a [&'a [u8]],
+        writable_desc_lengths: &'a [u32],
+    }
 
-        let desc_out = Descriptor::new(
-            next_addr,
-            size_of::<virtio_gpu_ctrl_hdr>() as u32,
-            VRING_DESC_F_NEXT as u16,
-            index + 1,
-        );
-        next_addr += desc_out.len() as u64;
-        index += 1;
+    fn gpu_backend_pair() -> (UnixStream, GpuBackend) {
+        let (frontend, backend) = UnixStream::pair().unwrap();
+        let backend = GpuBackend::from_stream(backend);
 
-        mem.write_obj::<virtio_gpu_ctrl_hdr>(ctrl_hdr, desc_out.addr())
-            .unwrap();
-        descriptors.push(desc_out);
-
-        // Buf descriptor: optional
-        if !buf.is_empty() {
-            let desc_buf = Descriptor::new(
-                next_addr,
-                buf.len() as u32,
-                (VRING_DESC_F_WRITE | VRING_DESC_F_NEXT) as u16,
-                index + 1,
-            );
-            next_addr += desc_buf.len() as u64;
+        (frontend, backend)
+    }
 
-            mem.write(buf, desc_buf.addr()).unwrap();
-            descriptors.push(desc_buf);
+    fn event_fd_into_file(event_fd: EventFd) -> File {
+        // SAFETY: safe because, the file descriptor is valid,
+        // We make sure to mem::forget the `event_fd` to not close it at scope exit of this function
```

Comment:
small nitpic - improve the safety documentation:
```// SAFETY: We ensure that the `event_fd` is properly handled such that its
// file descriptor is not closed while `File` takes ownership of it.```

Reply:
I changed it but replaced the word "while" with "after", that seem more clear to me.
```
 // SAFETY: We ensure that the `event_fd` is properly handled such that its file descriptor
 // is not closed after `File` takes ownership of it.
```